[Gradient Compression] Provide 2 default C++ comm hooks

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46959 [Gradient Compression] Surface C++ comm hooks to Python API as built-in comm hooks
* **#46701 [Gradient Compression] Provide 2 default C++ comm hooks**
* #46566 [Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.

Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

Commit:
[Gradient Compression] Provide 2 default C++ comm hooks

Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -0,0 +1,44 @@
+
+namespace c10d {
+
+c10::intrusive_ptr<torch::jit::Future> allReduceHook(
+    ProcessGroup* process_group,
+    GradBucket& bucket) {
+  auto allreduce_work = process_group->allreduce(*bucket.getTensorsRef());
+
+  auto div_by_process_group_size = [allreduce_work, process_group]() {
+    auto tensor = allreduce_work->result()[0] / process_group->getSize();
+    return c10::IValue(tensor);
+  };
+
+  auto fut = allreduce_work->getFuture();
+  return fut->then(div_by_process_group_size, fut->elementType());
+}
```
We probably need to have tests for these hooks. Ideally these should be python tests once we hook up the entire structure to python land.

SciPioneer replied:
I tested it locally by some hacking and reusing the same Python tests. The follow-up change is pybind the C++ implementation, and then we can add some tests.

rohan-varma commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -0,0 +1,44 @@
+  }
+  auto allreduce_work = process_group->allreduce(*tensors);
+
+  auto decompress_and_div_by_process_group_size = [allreduce_work,
```
Nit: Should we dedupe the `div_by_process_group` into a helper function?

SciPioneer replied:
There is a subtle difference between the divisions in allReduceHook and fp16CompressHook, as the latter decompresses the input tensor before division. Such decompression via `tensor.to(torch::kFloat)` will return a new temporary tensor that requires an in-place copy via `tensor.copy_`.

If we write a helper function for `div_by_process_group`, then it can only wrap up one line, which is dividing an input tensor by group size, but excluding `allreduce_work->result()[0]` and `c10::IValue(tensor)`.

SciPioneer commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -0,0 +1,44 @@
+
+  auto decompress_and_div_by_process_group_size = [allreduce_work,
+                                                   process_group]() {
+    auto tensor = allreduce_work->result()[0] / process_group->getSize();
```
Previously I intentionally applied the division on fp16 instead of fp32 to save some computation. After a second thought, I think it is probably safer to do the division on the higher-precision gradients.

SciPioneer commented on torch/csrc/distributed/c10d/default_comm_hooks.h:
```diff
@@ -0,0 +1,16 @@
+
+namespace c10d {
+
+c10::intrusive_ptr<torch::jit::Future> allReduceHook(
+    ProcessGroup* process_group,
+    GradBucket& bucket);
+
+c10::intrusive_ptr<torch::jit::Future> fp16CompressHook(
+    ProcessGroup* process_group,
+    GradBucket& bucket);
```
Had the same question too. My concern is the pybind part -- cannot construct such a subclass instance w/o providing a state, but can construct a callback/hook w/o such an input state, and the state is only needed when the callback/hook is invoked.

pritamdamania87 replied:
I don't think we need to provide state as part of pybind for this. On the python side the builtin hooks would look like this:

```
class BuiltinCommHook(Enum):
    ALLREDUCE = 1
    FP16_COMPRESS = 2

def register_builtin_comm_hook(builtin_hook: BuiltCommHook):
```

Then `register_builtin_comm_hook` will call an internal C++ method in reducer that is pybinded to Create the appropriate CppCommHook class and register it.

SciPioneer commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -31,6 +31,10 @@ class GradBucket {
     return tensors_;
   }
 
+  std::vector<at::Tensor>* getTensorsRef() {
+    return &tensors_;
+  }
```
getTensors is not used in C++ code, but there is a python binding to `get_tensors` of `dist._GradBucket`, and it is used in default_hooks.py.

I still want to keep the const semantics for that method, so added this one.

pritamdamania87 replied:
In that case can we change the return type to `std::vector<at::Tensor>&` instead of `std::vector<at::Tensor>*`? That is usually the convention we use.

Commit:
Update on "[Gradient Compression] Provide 2 default C++ comm hooks"


Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Provide 2 default C++ comm hooks"


Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Provide 2 default C++ comm hooks"


Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Provide 2 default C++ comm hooks"


Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Provide 2 default C++ comm hooks"


Provide 2 built-in implementations of C++ comm hook.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24382504](https://our.internmc.facebook.com/intern/diff/D24382504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24382504/)!

[ghstack-poisoned]
